src/Suave/Owin.fs

```diff
@@ -0,0 +1,621 @@
+      (fun v x -> x |> Map.put "principal" (box v))
+
+    let i2sc : Property<HttpCode, int> =
+      (fun x -> x.code),
+      (fun v x -> HttpCode.TryParse v |> Option.get) // NOTE: assumes user only sets valid codes
+
+    let owinMap ct requestHeadersLens responseHeadersLens responseStreamLens =
+      [ // 3.2.1 Request Data
+        // writeable / value???
+        OwinConstants.requestScheme,        HttpContext.request_ >--> HttpRequest.url_ >--> uriScheme <--> untyped
+        // writeable / value
+        OwinConstants.requestMethod,        req HttpRequest.method_
+        // writeable / value
+        OwinConstants.requestPathBase,      run HttpRuntime.homeDirectory_
+        // writeable / value
+        OwinConstants.requestPath,          HttpContext.request_ >--> HttpRequest.url_ >--> uriAbsolutePath <--> untyped
+        // writeable / value
+        OwinConstants.requestQueryString,   req HttpRequest.rawQuery_
+        // writeable / value???
+        OwinConstants.requestProtocol,      HttpContext.request_ >--> HttpRequest.httpVersion_ >--> hv2p <--> untyped
+        // !! mutation expected (!)
+        OwinConstants.requestHeaders,       HttpContext.request_ >--> HttpRequest.headers_ >--> requestHeadersLens <--> untyped
+        // writeable / value
+        OwinConstants.requestBody,          HttpContext.request_ >--> HttpRequest.rawForm_ >--> bytesToStream <--> untyped
+        OwinConstants.requestId,            HttpContext.request_ >--> HttpRequest.trace_ >--> Logging.TraceHeader.traceId_ >--> stringlyTyped uint64 <--> untyped
+        // writeable / value
+        OwinConstants.requestUser,          HttpContext.userState_ >--> claimsPrincipal <--> untyped
+
+        // 3.2.2 Response Data
+        // writeable / value
+        OwinConstants.responseStatusCode,   HttpContext.response_ >--> HttpResult.status_ >--> i2sc <--> untyped
+        // writeable / value
+        OwinConstants.responseReasonPhrase, HttpContext.response_ <--> untyped // TODO: add support for modifying phrasing to Core?
+        // writeable / value???
+        OwinConstants.responseProtocol,     HttpContext.request_ >--> HttpRequest.httpVersion_ >--> hv2p <--> untyped
+        // !! mutation expected
+        OwinConstants.responseHeaders,      HttpContext.response_ >--> HttpResult.headers_ >--> responseHeadersLens <--> untyped
+        // !! mutation expected
+        OwinConstants.responseBody,         HttpContext.response_ >--> HttpResult.content_ >--> responseStreamLens <--> untyped
+
+        // 3.2.3 Other Data
+        OwinConstants.callCancelled, ((fun x -> ct), (fun v x -> x)) <--> untyped // TODO: support cancellation token in HttpRequest signalling aborted request
+        OwinConstants.owinVersion, ((fun x -> "1.3"), (fun v x -> x)) <--> untyped
+
+        // per-request storage
+        "userData", HttpContext.userState_ <--> untyped
+      ]
+
+  type internal OwinDictionary(initialState) =
+    // TODO: support streaming optionally
```

Comment:
Do we support streaming in the first release? Does someone want to hack something as a continuation of this PR?